src/server/game/AuctionHouse/AuctionHouseListing.cpp

```diff
@@ -0,0 +1,598 @@
+/*
+ * Copyright (C) 2008-2017 TrinityCore <http://www.trinitycore.org/>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AuctionHouseListing.h"
+#include "AuctionHouseMgr.h"
+#include "Creature.h"
+#include "GameEventMgr.h"
+#include "Log.h"
+#include "ObjectAccessor.h"
+#include "ObjectMgr.h"
+#include "Player.h"
+#include "GameTime.h"
+
+#include <boost/thread/locks.hpp>
+#include <boost/thread/mutex.hpp>
+
+#include <vector>
+#include <unordered_map>
+
+namespace
+{
+    std::atomic<bool> _auctionHouseListingAllowed;
+
+    std::list<AuctionListingEvent*> _requestsNew;
+    std::list<AuctionListingEvent*> _requests;
+
+    boost::mutex _newRequestsLock;
+    boost::mutex _listingLock;
+
+    // Map of throttled players for GetAll, and throttle expiry time
+    // Stored here, rather than player object to maintain persistence after logout
+    std::unordered_map<ObjectGuid, time_t> GetAllThrottleMap;
+}
+
+void AuctionHouseListing::AuctionHouseListingThread()
+{
+    TC_LOG_INFO("misc", "Auction House Listing Thread starting");
+
+    while (!World::IsStopped())
+    {
+        if (IsListingAllowed())
+        {
+            boost::lock_guard<boost::mutex> lock(_listingLock);
+
+            // process new requests
+            {
+                boost::lock_guard<boost::mutex> lock(_newRequestsLock);
+                if (!_requestsNew.empty())
+                {
+                    std::move(_requestsNew.begin(), _requestsNew.end(), std::back_inserter(_requests));
+                    _requestsNew.clear();
+                }
+            }
+
+            if (!_requests.empty())
+            {
+                for (auto itr = _requests.begin(); itr != _requests.end();)
+                {
+                    if ((*itr)->Execute())
+                    {
+                        delete (*itr);
+                        _requests.erase(itr++);
+                    }
+                    else
+                        ++itr;
+
+                    // main thread may be waiting for us
+                    if (!IsListingAllowed())
+                        break;
+                }
+            }
+        }
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
```

Comment:
use `notification_variables` to properly wake up this thread rather than sleeping it.